### PR TITLE
DCOS-12774: [1/2] Fix multi-container health-check parsers & reducers

### PR DESCRIPTION
**Pull request by wavesoft**
This PR fixes the multi-container health-check parsers & reducers

### CODE REVIEW

**Comment by wavesoft** [comment]
retest this please

**Comment by nLight** [comment]
I think the ADD/REMOVE_ITEM solution should be copied here

**Comment by mlunoe** [comment]
Tested this out with the final PR https://github.com/dcos/dcos-ui/pull/1728, and it works well

File: plugins/services/src/js/reducers/serviceForm/MultiContainerHealthChecks.js [hunk 1]
```diff
@@ -1,81 +1,354 @@
...
+          break;
+      }
+
+      return newState;
+    }
+
+    // Assign properties
+    switch (group) {
+      case 'exec':
+        reduceCommandHealthCheck(newState, secondField, value);
```
Comment (mlunoe) [inline plugins/services/src/js/reducers/serviceForm/MultiContainerHealthChecks.js]: Comment from @nLight:
https://github.com/dcos/dcos-ui/pull/1728#discussion_r94931809
"I'd rather not mutate newState here
then we could even return reduceFormCommandHealthCheck(...) right away"

**Comment by wavesoft** [comment]
* Imported ADD_ITEM/REMOVE_ITEM solution here
* Added tests in order to be able to validate it
* Addressed the mutation issue

**Comment by orlandohohmeier** [comment]
**mesosphere-ci** retest this please 🙏

**Comment by wavesoft** [comment]
retest this please

File: plugins/services/src/js/reducers/serviceForm/MultiContainerHealthChecks.js [hunk 2]
```diff
@@ -3,79 +3,423 @@ import {
...
+    }
+
+    // Add an ADD_ITEM transaction
+    memo.push(new Transaction(path, null, ADD_ITEM));
+
+    // Parse detailed fields according to type
+    if (healthCheck.http != null) {
+      memo.push(new Transaction(path.concat(['protocol']), HTTP, SET));
+      memo.push(
+        ...parseHttpHealthCheck(healthCheck.http, path.concat(['http']), memo)
```
Comment (mlunoe) [inline plugins/services/src/js/reducers/serviceForm/MultiContainerHealthChecks.js]: #nit why not use `concat` for this? Shouldn't hold up the PR

Comment (wavesoft) [inline plugins/services/src/js/reducers/serviceForm/MultiContainerHealthChecks.js]: To avoid creating yet another array instance. But not really sure if that's helping...

Comment (nLight) [inline plugins/services/src/js/reducers/serviceForm/MultiContainerHealthChecks.js]: It doesn't really
http://babeljs.io/repl/#?babili=false&evaluate=true&lineWrap=false&presets=latest%2Creact%2Cstage-2&experimental=false&loose=false&spec=false&code=memo.push(%0A%20%20...parse(h.http%2C%20path.concat(%5B'http'%5D)%2C%20memo)%0A)%3B&playground=true

**Comment by wavesoft** [comment]
Let's get over with this 😄

**Comment by wavesoft** [comment]
run unit tests

File: plugins/services/src/js/reducers/serviceForm/MultiContainerHealthChecks.js [hunk 3]
```diff
@@ -3,79 +3,423 @@ import {
...
   SET
 } from '../../../../../../src/js/constants/TransactionTypes';
-import Util from '../../../../../../src/js/utils/Util';
+import {COMMAND, HTTP, HTTPS, TCP} from '../../constants/HealtCheckProtocols';
+import Transaction from '../../../../../../src/js/structs/Transaction';
+import ValidatorUtil from '../../../../../../src/js/utils/ValidatorUtil';
 
-function mapHealthChecks(item) {
-  const newItem = Util.omit(item, ['command', 'protocol']);
+function intOrNull(value) {
```
Comment (orlandohohmeier) [inline plugins/services/src/js/reducers/serviceForm/MultiContainerHealthChecks.js]: Any chance we can change this function to only "replace" `NaN` with `null`?  It kinda feels wrong to have a function that parses the input into an integer if the input isn't an empty string esp. as you can have strings (e.g. "e2") that still result in `NaN`.

File: plugins/services/src/js/reducers/serviceForm/Containers.js [hunk 4]
```diff
@@ -609,19 +560,15 @@ module.exports = {
...
+        this.healthCheckState[index] = {};
       }
 
-      newState[index].healthChecks = MultiContainerHealthChecks.call(
-        this.healthChecks[index],
-        newState[index].healthChecks,
-        {type, path: path.slice(2), value}
+      newState[index].healthCheck = multiContainerHealthFormReducer.call(
+        this.healthCheckState[index],
+        newState[index].healthCheck || {},
```
Comment (orlandohohmeier) [inline plugins/services/src/js/reducers/serviceForm/Containers.js]: Let's change `multiContainerHealthFormReducer` to handle undefined healthChecks instead, of defaulting the health check to an empty object here.

File: plugins/services/src/js/reducers/serviceForm/Containers.js [hunk 5]
```diff
@@ -456,19 +411,15 @@ module.exports = {
...
-    if (field === 'healthChecks') {
-      if (newState[index].healthChecks == null) {
-        newState[index].healthChecks = [];
-      }
-
-      if (this.healthChecks[index] == null) {
-        this.healthChecks[index] = {};
+    if (field === 'healthCheck') {
+      if (this.healthCheckState[index] == null) {
+        this.healthCheckState[index] = {};
```
Comment (orlandohohmeier) [inline plugins/services/src/js/reducers/serviceForm/Containers.js]: *Note to self: We need to refactor the container reducer as those indexes as well as the crazy number of conditions makes it really hard to follow through.*

File: plugins/services/src/js/reducers/serviceForm/MultiContainerHealthChecks.js [hunk 6]
```diff
@@ -3,79 +3,423 @@ import {
...
+ * @param {Array} path - The path prefix to the transaction
+ * @returns {Array} - Returns an array with the new transactions to append
+ */
+function parseCommandHealthCheck(healthCheck, path) {
+  const memo = [];
+  const {command={}} = healthCheck;
+
+  if (command.shell != null) {
+    memo.push(new Transaction(
+      path.concat(['command', 'shell']),
```
Comment (orlandohohmeier) [inline plugins/services/src/js/reducers/serviceForm/MultiContainerHealthChecks.js]: Let's rename this to `command.type` to make obvious that this is toggling the command type

File: plugins/services/src/js/reducers/serviceForm/MultiContainerHealthChecks.js [hunk 7]
```diff
@@ -3,79 +3,423 @@ import {
...
+      true,
+      SET
+    ));
+
+    memo.push(new Transaction(
+      path.concat(['command', 'string']),
```
Comment (orlandohohmeier) [inline plugins/services/src/js/reducers/serviceForm/MultiContainerHealthChecks.js]: Please rename this one to `command.value` to make this view agnostic.

File: plugins/services/src/js/reducers/serviceForm/Containers.js (hunk 5 above)
Comment (mlunoe) [inline plugins/services/src/js/reducers/serviceForm/Containers.js]: Yes, in general we need to split it up into multiple reducers, that we combine instead of this mega-file of logic

File: plugins/services/src/js/reducers/serviceForm/MultiContainerHealthChecks.js (hunk 6 above)
Comment (mlunoe) [inline plugins/services/src/js/reducers/serviceForm/MultiContainerHealthChecks.js]: Let's make sure that we are following the app configuration here, and don't something different. We've been burnt by this in the passed.

File: plugins/services/src/js/reducers/serviceForm/MultiContainerHealthChecks.js (hunk 7 above)
Comment (mlunoe) [inline plugins/services/src/js/reducers/serviceForm/MultiContainerHealthChecks.js]: Let's make sure that we are following the app configuration here, and don't something different. We've been burnt by this in the passed.

File: plugins/services/src/js/reducers/serviceForm/MultiContainerHealthChecks.js (hunk 6 above)
Comment (orlandohohmeier) [inline plugins/services/src/js/reducers/serviceForm/MultiContainerHealthChecks.js]: Where exactly do you see the problem/risk? The app and pod spec have a very different structure thus we shouldn't get burnt if the respective parses/reducers look different. If we want to align them, as initially planned, we need to finally move away from mirroring the app and pod spec JSON and find a better/abstract representation for the data. I'll create a JIRA to keep track of that.

File: plugins/services/src/js/reducers/serviceForm/MultiContainerHealthChecks.js (hunk 3 above)
Comment (wavesoft) [inline plugins/services/src/js/reducers/serviceForm/MultiContainerHealthChecks.js]: Good point. Also renamed it to `parseIntOrNull` for clarity.

File: plugins/services/src/js/reducers/serviceForm/MultiContainerHealthChecks.js (hunk 6 above)
Comment (wavesoft) [inline plugins/services/src/js/reducers/serviceForm/MultiContainerHealthChecks.js]: Renamed to `command.type`.

@mlunoe , I agree with @orlandohohmeier that whatever we keep in the batch should **never** be considered as an 1:1 representation of the data. The batch should contain whatever high-level data structures we might need on the UI in order to properly display and process the data. I also agree that we don't have a clean way of describing the data structures inside the batch, but this is something we have to address separately.

File: plugins/services/src/js/reducers/serviceForm/MultiContainerHealthChecks.js (hunk 7 above)
Comment (wavesoft) [inline plugins/services/src/js/reducers/serviceForm/MultiContainerHealthChecks.js]: Renamed to `command.value`

File: plugins/services/src/js/reducers/serviceForm/Containers.js (hunk 4 above)
Comment (wavesoft) [inline plugins/services/src/js/reducers/serviceForm/Containers.js]: Actually, it's already handled, so I will just remove the `|| {}`

**Comment by rcorral** [comment]
@orlandohohmeier could you review this again please?
